### PR TITLE
[ML] Identify multimodal prior distributions for anomaly explanation

### LINKED ISSUE #2416: [ML] Identify multimodal prior distributions for anomaly explanation
As an enhancement for explaining anomalies, it would be helpful to inform the users whether the prior distribution of the observed time series is multi-modal or has a single mode. For instance, if the distribution has two modes, then the _typical_ value may lie between the modes and hence not constitute the most probable value.

In particular, we can add a method
```c++
COneOfN::isMultimodal(){
  double sumOfWeights = ...
  for (const auto& model : m_Models){
    if(model.second.type() == E_Multimodal && model.first/sumOfWeights > 0.1) {
      return true;
    }
  }
  return false;
}
```

And then in the corresponding code in  `CtimSeriesModel`:
```c++
if (m_ResidualModel.type() == E_OneOfN) {
  return m_ResidualModel.isMultimodal();
}
return false;
```

### PR DESCRIPTION
**Pull request by valeriy42**
As an enhancement for explaining anomalies, it would be helpful to inform the users whether the prior distribution of the observed time series is multi-modal or has a single mode. For instance, if the distribution has two modes, the typical value may lie between the modes and hence not constitute the most probable value.

Closes #2416.

### CODE REVIEW

File: lib/maths/time_series/CTimeSeriesModel.cc [hunk 1]
```diff
@@ -1075,13 +1076,25 @@ bool CUnivariateTimeSeriesModel::uncorrelatedProbability(
...
     result.s_Probability = pOverall;
     result.s_FeatureProbabilities = std::move(featureProbabilities);
     result.s_Tail = {tail};
 
     return true;
 }
 
+bool CUnivariateTimeSeriesModel::multimodalPrior() const {
+    LOG_DEBUG(<< "Residual model type " << m_ResidualModel->type());
+    if (m_ResidualModel->type() == common::CPrior::E_OneOfN) {
```
Comment (tveasey) [inline lib/maths/time_series/CTimeSeriesModel.cc:1092]: So I can see why you did it like this. My inclination would be we have an `bool isSelectedModelMultimodal()` defined on `CPrior` which returns false and override it in `COneOfOneN` and ditch this function altogether.

**Review by tveasey — APPROVED** [review]
Just one real observation which relates to the implementation of the check to see if the selected residual model is multimodal. I have a couple of other very minor things I'll leave up to you. Happy for you to commit without another review.

File: lib/maths/common/COneOfNPrior.cc [hunk 2]
```diff
@@ -769,6 +769,19 @@ void COneOfNPrior::sampleMarginalLikelihood(std::size_t numberSamples,
...
+    double sumOfWeights{0.0};
+    for (const auto& weight : this->weights()) {
+        sumOfWeights += weight;
+    }
+    for (const auto& model : m_Models) {
+        if (model.second->type() == EPrior::E_Multimodal && model.first / sumOfWeights > 0.1) {
+            return true;
+        }
+    }
+    return false;
```
[the remark this answers is not in the document]
Comment (valeriy42) [reply lib/maths/common/COneOfNPrior.cc:782]: Thank you for this suggestion. 👍

File: lib/maths/time_series/CTimeSeriesModel.cc (hunk 1 above)
Comment (valeriy42) [reply lib/maths/time_series/CTimeSeriesModel.cc:1092]: I see how this can simplify the code. I updated the code according to your suggestion.

**Pull request merged**